esp-link/cgiwifi.c

```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+               apconf.beacon_interval,
+               apconf.ssid_hidden ? "enabled" : "disabled"
+               );
+    
+    jsonHeader(connData, 200);
+    httpdSend(connData, buff, -1);
     return HTTPD_CGI_DONE;
-  }
-
-  char url[64]; // redirect URL
-  if (os_strcmp(dhcp, "off") == 0) {
-    // parse static IP params
-    struct ip_info ipi;
-    bool ok = parse_ip(staticip, &ipi.ip);
-    if (nl > 0) ok = ok && parse_ip(netmask, &ipi.netmask);
-    else IP4_ADDR(&ipi.netmask, 255, 255, 255, 0);
-    if (gl > 0) ok = ok && parse_ip(gateway, &ipi.gw);
-    else ipi.gw.addr = 0;
-    if (!ok) {
-      jsonHeader(connData, 400);
-      httpdSend(connData, "Cannot parse static IP config", -1);
-      return HTTPD_CGI_DONE;
-    }
-    // save the params in flash
-    flashConfig.staticip = ipi.ip.addr;
-    flashConfig.netmask = ipi.netmask.addr;
-    flashConfig.gateway = ipi.gw.addr;
-    // construct redirect URL
-    os_sprintf(url, "{\"url\": \"http://%d.%d.%d.%d\"}", IP2STR(&ipi.ip));
-
-  } else {
-    // dynamic IP
-    flashConfig.staticip = 0;
-    os_sprintf(url, "{\"url\": \"http://%s\"}", flashConfig.hostname);
-  }
-
-  configSave(); // ignore error...
-  // schedule change-over
-  os_timer_disarm(&reassTimer);
-  os_timer_setfn(&reassTimer, configWifiIP, NULL);
-  os_timer_arm(&reassTimer, 1000, 0); // 1 second for the response of this request to make it
-  // return redirect info
-  jsonHeader(connData, 200);
-  httpdSend(connData, url, -1);
-  return HTTPD_CGI_DONE;
 }
 
+// ===== Wifi set mode end info functions
+
 //This cgi changes the operating mode: STA / AP / STA+AP
 int ICACHE_FLASH_ATTR cgiWiFiSetMode(HttpdConnData *connData) {
-  int len;
-  char buff[1024];
-
-  if (connData->conn==NULL) return HTTPD_CGI_DONE; // Connection aborted. Clean up.
-
-  len=httpdFindArg(connData->getArgs, "mode", buff, sizeof(buff));
-  if (len!=0) {
-    int m = atoi(buff);
-    DBG("Wifi switching to mode %d\n", m);
-    wifi_set_opmode(m&3);
-    if (m == 1) {
-      // STA-only mode, reset into STA+AP after a timeout if we don't get an IP address
-      os_timer_disarm(&resetTimer);
-      os_timer_setfn(&resetTimer, resetTimerCb, NULL);
-      os_timer_arm(&resetTimer, RESET_TIMEOUT, 0);
+    
+    if (connData->conn==NULL) return HTTPD_CGI_DONE; // Connection aborted. Clean up.
+    int len;
+    char buff[1024];
+    
+    int previous_mode = wifi_get_opmode();
+    
+    len=httpdFindArg(connData->getArgs, "mode", buff, sizeof(buff));
+    
+    int next_mode = atoi(buff);
+    
+    if (len!=0) {
+        if (next_mode == 2){
+            // moving to AP mode, so disconnect before leave STA mode
+            wifi_station_disconnect();
+        }
+        
+        DBG("Wifi switching to mode %d\n", next_mode);
+        wifi_set_opmode(next_mode&3);
+        
+        if (previous_mode == 2) {
+            // movint to STA-only mode from AP, so reset into STA+AP after a timeout if we don't get an IP address
```

Comment:
I have two different esp-link branch so I forgot to modify the original comments. 
I think actions are coherent:
if we are going to AP, disconnect STA
if we come from AP, we are going to STA or STA+AP, so a call to STA conf is needed
if we come from STA, we are going to AP or STA+AP, so a call to Soft-AP conf is needed